Add option to set Notifications Stream.

akhilmd opened the pull request:
Add option to set Notifications Stream from the administration settings.

Did not add a field to models.py. Added function to update realm's notification_stream based on stream name in actions.py. Added a new field in page_params to hold realm notification stream name. Added a dropdown in the administration page. Modified ```populate_db.py``` to have a notificatons_stream during the tests. Altered ```test_signup``` and ```test_subs```.

Fixes: #3708

smarx commented:
*Automated message from Dropbox CLA bot*

@akhilmd, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -176,7 +176,7 @@ def test_register(self):
             self.register("test@zulip.com", "test")
         # Ensure the number of queries we make is not O(streams)
-        self.assert_max_length(queries, 69)
+        self.assert_max_length(queries, 88)
```
Why did you change this?  It seems likely that adding 19 new database queries would be caused by a bug in your changes...

timabbott commented on zerver/lib/actions.py:
```diff
@@ -544,6 +544,21 @@ def do_set_realm_waiting_period_threshold(realm, threshold):
+def do_set_realm_notifications_stream(realm, stream_name):
+    # type: (Realm, Text) -> None
+
+    stream = get_stream(stream_name, realm)
```
this function should be passed a stream from its caller, rather than calling `get_stream`.  Our convention is functions with a `do_foo` name should not be able to fail, and this can fail if the stream name is invalid.  

(The caller should use `access_stream`, not `get_stream`, as well, for proper security checks and error handling)

timabbott commented on zerver/views/realm.py:
```diff
@@ -93,4 +95,7 @@ def update_realm(request, user_profile, name=REQ(validator=check_string, default
         data['waiting_period_threshold'] = waiting_period_threshold
+    if notifications_stream_name is not None and realm.notifications_stream.name != notifications_stream_name:
+        do_set_realm_notifications_stream(realm, notifications_stream_name)
+        data['notifications_stream_name'] = notifications_stream_name
```
you should add test logic to `test_realm.py` for this feature, similar to what's there, to give this code test coverage.

Also use `access_stream` to fetch the stream here.

timabbott commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -130,6 +131,14 @@
+        <option value='{{this.name}}'>{{this.name}}</option>
+        {{/each}}
+      </select>
+    </div>
```
UI-wise, I think we should probably have a "Change" link that opens a modal here, rather than having the selection menu in the page, since a dropdown menu is kinda cluttered

timabbott commented:
@akhilmd if you have time in the near future, it'd be great to rebase this (a ton of merge conflicts have shown up) so we can get it merged and stop growing new conflicts :).

timabbott commented:
@akhilmd just checking in on whether you've had a chance to update this!  It'd be great to get this merged.

timabbott commented:
It might be a bit easier to just redo the changes than do an actual rebase.

akhilmd commented on zilencer/management/commands/populate_db.py:
```diff
@@ -162,6 +162,8 @@ def handle(self, **options):
 
             bulk_create_streams(zulip_realm, stream_dict)
+            zulip_realm.notifications_stream = get_stream("Rome", zulip_realm)
+            zulip_realm.save(update_fields=['notifications_stream'])
```
I made this change based on [this](https://chat.zulip.org/#narrow/stream/issues/subject/.233708/near/157484) discussion. Backend tests fail due to test realm not having a `Realm.notifications_stream` set up.

akhilmd commented on zerver/tests/test_signup.py:
```diff
@@ -266,7 +266,7 @@ def test_register(self):
             self.register("test@zulip.com", "test")
         # Ensure the number of queries we make is not O(streams)
-        self.assert_length(queries, 47)
+        self.assert_length(queries, 70)
```
Looks like [this](https://github.com/zulip/zulip/pull/4156/commits/f1be02a4d965ee2bad7d0247f2d1ca7306cfdc11#r118293609) change is causing so many new database queries to be added. With only the `notifications_stream` addition in `populate_db.py` to master, this test fails.

akhilmd commented:
@timabbott sorry for the huge delay in getting this through!

I have addressed all your comments except the one on `test_signup.py`, for which I added a few comments.

I replaced the native dropdown with a bootstrap dropdown since it was possible to put a text field for filtering streams to avoid clutter as shown below:
![notif-stream-dropup](https://cloud.githubusercontent.com/assets/19967087/26413960/476adbfa-40cb-11e7-9363-6435e1c17736.gif)

umairwaheed commented on zerver/tests/test_signup.py:
```diff
@@ -267,7 +267,7 @@ def test_register(self):
             self.register(self.nonreg_email('test'), "test")
         # Ensure the number of queries we make is not O(streams)
-        self.assert_length(queries, 47)
+        self.assert_length(queries, 70)
```
How the number of queries increased by so much?

akhilmd replied:
[This](https://github.com/zulip/zulip/pull/4156/#discussion_r118293609) addition to populate_db.py to set a `realm.notifications_stream` is causing that increase.

timabbott replied:
I think probably that suggests some other bug that we should investigate.

timabbott commented:
Hi @akhilmd!  We have recently refactored the way most realm settings are done to use the new `property_types` framework, which eliminates a lot of the code one would need to add when adding a new realm setting, both on the frontend and backend.  See http://zulip.readthedocs.io/en/latest/new-feature-tutorial.html for notes on how this works today.  Can you update this to use that model?  I think you'll be able to eliminate most of the code in this PR and have things be nice and clean...

(Also, you should get the tests passing).

akhilmd commented:
@timabbott, I actually referred to that docs page continually while working on this. But, [this](http://zulip.readthedocs.io/en/latest/new-feature-tutorial.html#update-the-model) part says:
> When creating a realm property that is **not a boolean, Text or integer field**, or when adding a field that is dependent on other fields, handle it separately and do not add the field to the `property_types` dictionary.

Also, in `do_set_realm_property`, the `event[value]` should be JSONable. Here, a `Stream` object is not.

Therefore, I decided to not go with the new `property_types` framework.

HarshitOnGitHub commented on zerver/tests/test_realm.py:
```diff
@@ -152,6 +152,27 @@ def test_do_deactivate_realm_on_deactived_realm(self):
+        # type: () -> None
+        new_notif_stream_id = 4
+        realm = get_realm('zulip')
+        # we need an admin user.
```
Minor but I think this call to get_realm can be removed. Also capitalize the starting word of comment.

HarshitOnGitHub commented on zerver/lib/actions.py:
```diff
@@ -491,6 +491,19 @@ def do_set_realm_message_editing(realm, allow_message_editing, message_content_e
 
+def do_set_realm_notifications_stream(realm, stream):
+    # type: (Realm, Stream) -> None
+
```
Minor but this blank line should be removed.

timabbott commented:
> It would be good to break it into smaller commits, basically one for adding setting in the backend and other for adding the UI.

I actually don't think there's a need to do that split; while we could reasonably merge the backend without the frontend.  But if we think the backend is done and the frontend has a ways to go, that could be useful to do just to shrink the size of what's in review.

akhilmd commented:
@timabbott one thing I'm not quite sure about is whether the UI should allow the user to _deselect_ the current notifications stream. In which case, a stream-id of _-1_ can be sent to the backend and `Realm.notifications_stream` set to None.

timabbott commented:
Yep, we should support that, and that seems like a good approach from an API perspective.

timabbott commented on frontend_tests/casper_tests/10-admin.js:
```diff
@@ -19,6 +19,28 @@ casper.waitForSelector('#settings_overlay_container.show', function () {
+casper.waitUntilVisible('#admin-realm-notifications-stream-status', function () {
+    casper.test.assertSelectorHasText('#admin-realm-notifications-stream-status',
+                                      'Notifications stream changed!');
+});
```
These blocks should all be wrapped inside a `casper.then()` or this test will have race issues.

timabbott commented on static/js/settings_org.js:
```diff
@@ -103,6 +103,42 @@ exports.populate_auth_methods = function (auth_methods) {
+        realm_notif_stream_label.text(i18n.t("Current notifications stream: __stream_name__",
+                                             {stream_name: currrent_notif_stream.name}));
+    } else {
+        realm_notif_stream_label.text(i18n.t("Notifications stream not selected!"));
```
Make this "Notifications stream: disabled." ?

timabbott commented on static/js/settings_org.js:
```diff
@@ -103,6 +103,42 @@ exports.populate_auth_methods = function (auth_methods) {
+    var currrent_notif_stream = stream_data.get_sub_by_id(stream_id);
+    if (currrent_notif_stream) {
+        realm_notif_stream_label.text(i18n.t("Current notifications stream: __stream_name__",
+                                             {stream_name: currrent_notif_stream.name}));
```
Let's prefix with a `#` using `#__stream_name__`?

timabbott commented on static/js/settings_org.js:
```diff
@@ -103,6 +103,42 @@ exports.populate_auth_methods = function (auth_methods) {
+    $(".dropdown-search").click(function (e) {
+        e.stopPropagation();
+    });
+};
```
Any input in the search dropdown should be being cleared if you exit and return later, but isn't.

timabbott commented on static/styles/settings.css:
```diff
@@ -996,6 +996,18 @@ input[type=checkbox].inline-block {
+    overflow-y: auto;
+    margin-top: 0;
+    display: block;
+}
```
these CSS rules should be encapsulated inside `#id_realm_notifications_stream` so that there's no risk that they impact other parts of the app.

timabbott commented on static/templates/settings/admin-realm-dropdown-stream-list.handlebars:
```diff
@@ -0,0 +1,3 @@
+{{#with stream}}
+<li title="{{name}} has {{subscriber_count}} subscribers" class="stream_name" data-stream-id="{{stream_id}}"><a tabindex="-1">{{name}}</a></li>
```
can you line-wrap this more nicely?

We can probably remove the name from the title. and just say `{{subscriber_count}} subscribers`.

Well, except we should also tag the title for translation.

timabbott commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -18,6 +19,20 @@
+                    title="{{t 'The stream to which new stream notifications go to.' }}">
+                </label>
+                {{#if is_admin }}
+                <span name="realm_notifications_stream" aria-labelledby="realm_notifications_stream_label" class="dropup actual-dropdown-menu" id="id_realm_notifications_stream" data-selected-stream-id={{realm_notifications_stream_id}}>
```
please wrap this long line.

timabbott commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -18,6 +19,20 @@
+                <span name="realm_notifications_stream" aria-labelledby="realm_notifications_stream_label" class="dropup actual-dropdown-menu" id="id_realm_notifications_stream" data-selected-stream-id={{realm_notifications_stream_id}}>
+                    <a class="dropdown-toggle" data-toggle="dropdown">{{t "[Change]" }}</a>
+                    <ul class="dropdown-menu" role="menu">
+                        <li class="dropdown-search"><input tabindex="-1" type="text" placeholder="{{t 'Filter streams' }}" autofocus/></li>
```
put the `<input>` on its own line here.

timabbott commented on zerver/tests/test_realm.py:
```diff
@@ -152,6 +152,26 @@ def test_do_deactivate_realm_on_deactived_realm(self):
+        result = self.client_patch('/json/realm', req)
+        self.assert_json_error(result, 'Invalid stream id')
+        realm = get_realm('zulip')
+        self.assertNotEqual(realm.notifications_stream.id, invalid_notif_stream_id)
```
you should add a `test_events` test for this, as well.

akhilmd commented:
@timabbott, I updated as per reviews. Also, admins can disable notifications stream:
![tozu](https://user-images.githubusercontent.com/19967087/26991500-33a8ba56-4d78-11e7-9add-4db361116954.png)

timabbott commented:
@akhilmd it looks like you're failing the capitalization tests?

timabbott commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -19,6 +20,27 @@
             <div class="input-group">
+                <label for="realm_notifications_stream" id="realm_notifications_stream_label" class="inline-block"
+                    title="{{t 'The stream to which new stream notifications go to.' }}">
+                    {{t "Notifications Stream: " }}
```
We only capitalize the first letter in sentences (hence the linter error)

timabbott commented on static/js/settings_org.js:
```diff
@@ -107,6 +107,46 @@ exports.populate_auth_methods = function (auth_methods) {
     }
 };
 
+var realm_notif_stream;
```
we don't generally use abbreviations in Zulip; can you spell this one out?

timabbott commented on static/js/settings_org.js:
```diff
@@ -107,6 +107,46 @@ exports.populate_auth_methods = function (auth_methods) {
+    var currrent_notif_stream = stream_data.get_sub_by_id(stream_id);
+    if (currrent_notif_stream) {
+        $("#realm_notifications_stream_name").text(i18n.t("#__stream_name__",
+                                             {stream_name: currrent_notif_stream.name})).removeClass("text-warning");
```
this indentation seems way too much

timabbott commented on static/js/settings_org.js:
```diff
@@ -107,6 +107,46 @@ exports.populate_auth_methods = function (auth_methods) {
+                return item.name.toLowerCase().match(value);
+            },
+        },
+    }).init();
```
nice use of `list_render`.

timabbott commented on static/js/settings_org.js:
```diff
@@ -107,6 +107,46 @@ exports.populate_auth_methods = function (auth_methods) {
 
+var realm_notif_stream;
+
+exports.update_notifications_stream = function (stream_id) {
```
maybe this should be spelled `render_notifications_stream_ui`, for clarity?  This sounds like it might send something to the server.

timabbott commented on static/js/server_events_dispatch.js:
```diff
@@ -86,6 +86,8 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
                 settings_org.populate_auth_methods(event.data.authentication_methods);
             }
+        } else if (event.op === 'update' && event.property === 'notifications_stream_id') {
+            page_params.realm_notifications_stream_id = event.value;
```
shouldn't this also call the function to render the notifications stream UI?  That way, changes take effect in all browsers, not just the current one.

timabbott commented on static/templates/settings/organization-settings-admin.handlebars:
```diff
@@ -19,6 +20,27 @@
+                <a class="notifications-stream-disable">{{t "[Disable]" }}</a>
+                {{/if}}
+            </div>
+            <div class="input-group">
```
I think we should probably move this setting to live below the "Save changes" button and instead just update the server as soon as user selects something.  It's pretty confusing as-is; I'd never remember to click the "Save changes" button.

(As a general note, the plan is to get rid of that button)

akhilmd commented:
@timabbott I updated the PR so that on *selecting a stream from the dropdown menu* or on *clicking on Disable*, a patch request is sent to `/json/realm` with the appropriate stream-id, without the need to submit the form.

timabbott commented:
Merged, thanks @akhilmd!!  Congrats on finishing a big project.

I made one change before merging, which was removing the trailing whitespace in one of the translation strings.

timabbott commented:
Oh, a couple follow-up things to check:
* Make sure that looking at the page looks correct for a non-admin user.  In theory, it shouldn't show either the "Disable" or "change" buttons if you're not a realm admin.  @abhijeetkaur will have advice on how to make this work.
* We should document this in the /help/ docs.  See http://zulip.readthedocs.io/en/latest/user-docs.html for details.
